File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/projecting/StructStartSeekingProjector.java
package com.codepoetics.raffia.streaming.projecting;

import com.codepoetics.raffia.paths.Path;
import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

public final class StructStartSeekingProjector<T extends BasketWriter<T>> extends StreamingProjector<T> {

  private final Path path;

  public StructStartSeekingProjector(T target, Path path) {
    super(target, null);
    this.path = path;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    return StreamingProjector.startObject(target, path, this);
  }

  @Override
  public FilteringWriter<T> beginArray() {
    return StreamingProjector.startArray(target, path, this);
  }

  @Override
  public FilteringWriter<T> end() {
    throw new IllegalStateException("end() called when not writing object or array");
  }

  @Override
  public FilteringWriter<T> key(String key) {
    throw new IllegalStateException("key() called when not writing object");
  }

  @Override
  public FilteringWriter<T> add(String value) {
    return new EndOfLineProjector<>(target);
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    return new EndOfLineProjector<>(target);
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    return new EndOfLineProjector<>(target);
  }

  @Override
  public FilteringWriter<T> addNull() {
    return new EndOfLineProjector<>(target);
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/projecting/StructStartSeekingProjector.kt
package com.codepoetics.raffia.streaming.projecting

import com.codepoetics.raffia.paths.Path
import com.codepoetics.raffia.streaming.EndOfLineFilter
import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

internal class StructStartSeekingProjector<T : BasketWriter<T>>(target: T, private val path: Path) : StreamingProjector<T>(target, null) {

    override fun beginObject(): FilteringWriter<T> {
        return StreamingProjector.startObject(target, path, this)
    }

    override fun beginArray(): FilteringWriter<T> {
        return StreamingProjector.startArray(target, path, this)
    }

    override fun end(): FilteringWriter<T> {
        throw IllegalStateException("end() called when not writing object or array")
    }

    override fun key(key: String): FilteringWriter<T> {
        throw IllegalStateException("key() called when not writing object")
    }

    override fun add(value: String): FilteringWriter<T> {
        return EndOfLineFilter(target)
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        return EndOfLineFilter(target)
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        return EndOfLineFilter(target)
    }

    override fun addNull(): FilteringWriter<T> {
        return EndOfLineFilter(target)
    }
}
